lib/Sema/CSSimplify.cpp

```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+      }
+      if (tmp.keywordArgumentsMethod) {
+        if (methods.keywordArgumentsMethod &&
+            methods.keywordArgumentsMethod != tmp.keywordArgumentsMethod) {
+          error = true;
+          return methods;
+        }
+        methods.keywordArgumentsMethod = tmp.keywordArgumentsMethod;
+      }
+    }
+    return methods;
+  };
+
+  // Calculate @dynamicCallable methods.
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have the attribute.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
+      SmallVector<Type, 2> componentTypes;
+      for (auto protocolDecl : archetype->getConformsTo())
+        componentTypes.push_back(protocolDecl->getDeclaredType());
+      if (auto superclass = archetype->getSuperclass())
+        componentTypes.push_back(superclass);
+      return calculateForComponentTypes(componentTypes);
+    }
+
+    // If this is a protocol composition, check if any of its members have the
+    // attribute.
+    if (auto protocolComp = dyn_cast<ProtocolCompositionType>(canType))
+      return calculateForComponentTypes(protocolComp->getMembers());
+
+    // Otherwise, this must be a nominal type.
+    // Dynamic calling doesn't work for tuples, etc.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return DynamicCallableMethods();
+
+    // If this type conforms to a protocol which has the attribute, then
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
+
+    // Walk superclasses, if present.
+    llvm::SmallPtrSet<const NominalTypeDecl*, 8> visitedDecls;
+    while (1) {
+      // If we found a circular parent class chain, reject this.
+      if (!visitedDecls.insert(nominal).second)
+        return DynamicCallableMethods();
+
+      // If this type has the attribute on it, then look up the methods.
+      if (nominal->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
+
+      // If this type is a class with a superclass, check superclasses.
+      if (auto *cd = dyn_cast<ClassDecl>(nominal)) {
+        if (auto superClass = cd->getSuperclassDecl()) {
+          nominal = superClass;
+          continue;
+        }
+      }
+
+      return DynamicCallableMethods();
+    }
+  };
+
+  return CS.DynamicCallableCache[canType] = calculate();
+}
+
+// Simplify applicable function constraint for an application of a
+// @dynamicCallable type.
+static ConstraintSystem::SolutionKind
+simplifyDynamicCallableApplicableFnConstraint(
+    ConstraintSystem &CS, Type type1, Type type2,
+    ConstraintLocatorBuilder locator,
+    ConstraintSystem::TypeMatchOptions flags) {
+  using SolutionKind = ConstraintSystem::SolutionKind;
```

Comment:
This really doesn't behave like normal overload resolution. For example we fail to type-check the following, which I believe a user might normally expect to resolve to the overload with keyword arguments:

```
@dynamicCallable
struct S {
  func dynamicallyCall(withArguments: [String]) -> Int {
    fatalError()
  }
  func dynamicallyCall(withKeywordArguments: [String : String]) -> Float {
    fatalError()
  }
}

let s = S()
let _: Float = s("hi")
```

Reply:
This is a good point. I describe the problems with the lookup-based approach in more detail [here](https://github.com/apple/swift/pull/16980#issuecomment-410142870). Do you have ideas for a better method resolution approach?